msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AcquireTokenByManagedIdentitySupplier.java

```diff
@@ -93,15 +95,27 @@ private AuthenticationResult fetchNewAccessTokenAndSaveToCache(TokenRequestExecu
     }
 
     private AuthenticationResult createFromManagedIdentityResponse(ManagedIdentityResponse managedIdentityResponse) {
-        long expiresOn = Long.valueOf(managedIdentityResponse.expiresOn);
-        long refreshOn = expiresOn > 2 * 3600 ? (expiresOn / 2) : 0L;
+        long expiresOn = Long.parseLong(managedIdentityResponse.expiresOn);
+        long refreshOn = calculateRefreshOn(expiresOn);
+        AuthenticationResultMetadata metadata = AuthenticationResultMetadata.builder()
+                .refreshOn(refreshOn)
+                .build();
 
         return AuthenticationResult.builder()
                 .accessToken(managedIdentityResponse.getAccessToken())
                 .scopes(managedIdentityParameters.resource())
                 .expiresOn(expiresOn)
                 .extExpiresOn(0)
                 .refreshOn(refreshOn)
+                .metadata(metadata)
                 .build();
     }
+
+    private long calculateRefreshOn(long expiresOn){
+        long timestampSeconds = System.currentTimeMillis() / 1000;
+        long expiresIn = expiresOn - timestampSeconds;
+
+        //The refreshOn value should be half the value of the token lifetime, if the lifetime is greater than two hours
+        return expiresIn > TWO_HOURS ? (expiresIn / 2) + timestampSeconds : 0;
```

Comment:
Is there a difference between 0 and NULL (absence of a value) for refreshOn? 

In other words, is there a danger that MSAL interprets refreshOn = 0 as "always refresh" if expiry < 2h ?

Note: In MSAL.NET the implementation differentiates between NULL and int (0 or non-0). Not suggesting you do the same, I think it's fair to say that when `refreshOn = 0` or some other sentinel value, then you ignore it. Just want to make sure this is really the case.

Reply:
If you look at this part of AcquireTokenSilentSupplier you'll see one of the places where this value gets used internally: https://github.com/AzureAD/microsoft-authentication-library-for-java/blob/310cc07d912acc6d4e3371e569ba660c916f98fc/msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AcquireTokenSilentSupplier.java#L138

In that if statement 'null' and '0' are treated the same, if refreshon is either value then we won't use it as a reason to refresh.

Reply:
Thanks for confirming!